File: providers/aws/src/test/java/cloudspec/aws/ec2/EC2LocalGatewayLoaderTest.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2;

import cloudspec.aws.ec2.loader.EC2LocalGatewayLoader;
import datagen.services.ec2.model.LocalGatewayGenerator;
import org.junit.Test;
import software.amazon.awssdk.services.ec2.model.DescribeLocalGatewaysResponse;
import software.amazon.awssdk.services.ec2.model.LocalGateway;

import java.util.function.Consumer;

import static org.junit.Assert.*;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.when;

public class EC2LocalGatewayLoaderTest extends EC2LoaderTest {
    private final EC2LocalGatewayLoader loader = new EC2LocalGatewayLoader(clientsProvider);

    @Test
    public void shouldLoadLocalGateways() {
        var localGateways = LocalGatewayGenerator.localGateways();
        when(ec2Client.describeLocalGateways(any(Consumer.class))).thenReturn(
                DescribeLocalGatewaysResponse.builder()
                                                .localGateways(localGateways.toArray(new LocalGateway[0]))
                                                .build()
        );

        var resources = loader.getAll();

        assertNotNull(resources);
        assertEquals(localGateways.size(), resources.size());
        assertTrue(localGateways.containsAll(resources));
    }

    @Test
    public void shouldNotLoadLocalGatewayByRandomId() {
        when(ec2Client.describeLocalGateways(any(Consumer.class))).thenReturn(
                DescribeLocalGatewaysResponse.builder()
                                                .build()
        );

        var resourceOpt = loader.getById(LocalGatewayGenerator.localGatewayId());

        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isEmpty());
    }

    @Test
    public void shouldLoadLocalGatewayById() {
        var localGateways = LocalGatewayGenerator.localGateways(1);
        when(ec2Client.describeLocalGateways(any(Consumer.class))).thenReturn(
                DescribeLocalGatewaysResponse.builder()
                                                .localGateways(localGateways.toArray(new LocalGateway[0]))
                                                .build()
        );

        var localGateway = localGateways.get(0);
        var resourceOpt = loader.getById(localGateway.localGatewayId());
        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isPresent());
        assertEquals(resourceOpt.get(), localGateway);
    }
}

File: providers/aws/src/test/kotlin/cloudspec/aws/ec2/EC2LocalGatewayLoaderTest.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import arrow.core.None
import arrow.core.Some
import datagen.services.ec2.model.LocalGatewayGenerator.localGatewayId
import datagen.services.ec2.model.LocalGatewayGenerator.localGateways
import io.mockk.every
import org.junit.Test
import software.amazon.awssdk.services.ec2.model.DescribeLocalGatewaysRequest
import software.amazon.awssdk.services.ec2.model.DescribeLocalGatewaysResponse
import java.util.function.Consumer
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class EC2LocalGatewayLoaderTest : EC2LoaderTest() {
    private val loader = EC2LocalGatewayLoader(clientsProvider)

    @Test
    fun `should load all resources`() {
        val localGateways = localGateways()

        every {
            ec2Client.describeLocalGateways(any<Consumer<DescribeLocalGatewaysRequest.Builder>>())
        } answers {
            val builder = DescribeLocalGatewaysRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeLocalGatewaysRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertTrue(request.filters().isEmpty())

            DescribeLocalGatewaysResponse.builder()
                    .localGateways(localGateways)
                    .build()
        }

        val resources = loader.all(emptyList())
        assertNotNull(resources)
        assertEquals(localGateways.size, resources.unsafeRunSync().size)
    }

    @Test
    fun `should not load resource by random id`() {
        every {
            ec2Client.describeLocalGateways(any<Consumer<DescribeLocalGatewaysRequest.Builder>>())
        } answers {
            val builder = DescribeLocalGatewaysRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeLocalGatewaysRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeLocalGatewaysResponse.builder()
                    .build()
        }

        val resourceOpt = loader.byId(emptyList(), localGatewayId())
        assertTrue(resourceOpt.unsafeRunSync() is None)
    }

    @Test
    fun `should load resource by id`() {
        val localGateways = localGateways()

        every {
            ec2Client.describeLocalGateways(any<Consumer<DescribeLocalGatewaysRequest.Builder>>())
        } answers {
            val builder = DescribeLocalGatewaysRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeLocalGatewaysRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeLocalGatewaysResponse.builder()
                    .localGateways(localGateways)
                    .build()
        }

        val localGateway = localGateways[0]
        val resourceOpt = loader.byId(emptyList(), localGateway.localGatewayId())
        assertTrue(resourceOpt.unsafeRunSync() is Some<*>)
    }
}
